### PR TITLE
fix: tell an unscanned week apart from an unknown one in token activity

### LINKED ISSUE #2893: Token activity "Cumulative" view renders entirely blank at the default 30-day cost history setting
# Token activity "Cumulative" view renders entirely blank at the default 30-day cost history setting

## Summary

In Settings > Usage & Spend, the Token activity card offers Daily, Weekly, and Cumulative modes. The Cumulative mode paints all 53 columns with the "Unavailable" fill. The card shows no data at all.

The cause is a running AND over the weekly coverage flags. `SpendActivityAggregateSeries.cumulative()` marks a week as covered only when every earlier week is also covered. The heatmap spans 365 days, but `tokenCostUsageHistoryDays` defaults to 30. The first week of the grid is therefore always uncovered, the flag turns false at index 0, and it never turns true again.

This affects every user who keeps the default history setting. Daily and Weekly work correctly, because both evaluate coverage per cell.

## Environment

- CodexBar: 0.49.3 (build 117)
- macOS: 26.6.1
- Providers with local cost history: Claude and Codex
- `tokenCostUsageHistoryDays`: 30 (default, not overridden)
- Card header reads: `4.2B · Coverage: 30 / 365`

## Steps to reproduce

1. Enable token cost usage, and keep the default 30-day history setting.
2. Open Settings > Usage & Spend.
3. Scroll to the Token activity card.
4. Select Daily. The grid shows recent activity in green.
5. Select Weekly. The grid shows recent weeks in green.
6. Select Cumulative. Every column turns gray.

## Expected and actual

Expected: a running total that rises from left to right, and reaches full column height in the most recent week.

Actual: 53 gray columns. The caption reads "Running total", and the legend on the right shows the "Unavailable" swatch.

## Root cause 1: the running AND

`Sources/CodexBar/SpendActivityHeatmap.swift:139`

```swift
coverageIsComplete = coverageIsComplete && self.isCovered[index]
```

`SpendActivityWeekGrid` paints every row of an uncovered column with `SpendActivityLevels.unavailableFill`. That fill is very close to the level-0 fill, so the whole grid reads as empty rather than as an error.

The behavior is deliberate, and a test locks it in at `Tests/CodexBarTests/SpendActivityHeatmapTests.swift:196`. The intent is defensible. A running total after a gap is a lower bound, not a fact. The result is still a view that can never show anything at the default setting.

## Root cause 2: the coverage window is 30 days while the grid spans 365 days

`SpendDashboardModel.tokenActivity` marks a day unknown when any single provider fails to cover it (`Sources/CodexBar/SpendDashboardModel.swift:472`). Coverage per provider comes from `sourceCoverageInterval`, which walks back `snapshot.historyDays` from the snapshot timestamp (`Sources/CodexBar/SpendDashboardModel.swift:586`).

Codex accounts receive a second scan at 365 days for the heatmap (`Sources/CodexBar/SpendDashboardController.swift:392-399`, `activityDays = 365`). Every other provider reaches the dashboard with the snapshot that `UsageStore` published, and that scan uses `settings.costUsageHistoryDays`. The default is 30 (`Sources/CodexBar/SettingsStore.swift:482`).

So a Claude plus Codex user gets a 30-day intersection, which matches the `Coverage: 30 / 365` in the header. The 365-day heatmap then has 335 unknown days by construction.

Inferred, not yet verified at runtime: raising the history setting to 365 should widen coverage to 365 of 365, and should make Cumulative render. If that is correct, the blank view is a default-configuration bug, which explains the absence of earlier reports.

## Secondary issue: partial weeks drop covered tokens

`weeklyActivity()` marks a week covered only when all visible days are covered (`Sources/CodexBar/SpendActivityHeatmap.swift:112`). The oldest covered week is almost always partial, because the 30-day window rarely lands on a week boundary.

In this report the week of 2026-07-12 holds about 1.0B covered tokens. Weekly and Cumulative both drop it. The Weekly grid shows 4 green columns instead of 5, and the sum of the visible columns does not reconcile with the 4.2B in the card header.

## Suggested fixes

1. Start the running total and the coverage flag at the first covered week. Keep the leading unknown weeks gray. Label the total as a lower bound while coverage is partial.
2. Give every cost-capable provider the same 365-day activity scan that Codex already receives, or decouple the heatmap window from `costUsageHistoryDays`.
3. For a partially covered week, either render the partial total with a distinct treatment, or exclude its days from the card total, so that the visible columns reconcile with the header.

Item 1 alone removes the blank grid. Item 2 is needed before the view carries useful information.

## Data point

Local Claude and Codex logs on this machine hold about 7.5B tokens over 176 days, back to 2025-11-11. Rendered as a running total, the curve stays flat through the winter and then rises steeply from June onward. None of that shape is visible today.

### PR DESCRIPTION
**Pull request by urda**
## Summary

The Cumulative mode of the Token activity card renders entirely blank. All 53 columns use the
unavailable fill, and every hover reports "Unavailable".

The root cause is that one flag stood in for two different facts.
`TokenActivityInputSummary.tokens(on:)` returns nil when a day falls outside `coveredInterval`,
when the provider is globally invalid, when the day is in `invalidDays`, and when the day has no
entry while `hasCompleteHistory` is false. Only the first means "never scanned". By the time the
heatmap saw the data, all four had collapsed into `isCovered == false`.

The grid always spans 365 days, while the scan window comes from `tokenCostUsageHistoryDays`, which
defaults to 30. So the first week of the grid is unscanned for every user on the default setting,
and the running AND in `cumulative()` turned false at index 0 and never recovered.

This change carries scan provenance down to the heatmap. `TokenActivityPoint` gains `isScanned`,
`SpendActivitySeries` carries it per day, and `SpendActivityAggregateSeries` carries it per week.
Three states now exist where there were two: not scanned, scanned but unknown, and covered.

- The running total starts at the first **scanned** week, so an unscanned prefix no longer blanks
  the year.
- A week that was scanned but could not be resolved is still a real gap, so every later total stays
  unavailable, because it is only a lower bound.
- A week is covered when every **scanned** day inside it is covered. The week straddling the start
  of the scan window is therefore available for its scanned portion instead of being discarded.

Totals are untouched. The accumulator still runs on every iteration, so the last column continues
to reconcile with the total in the card header.

Fixes #2893.

## Proof

Patched build, `History window` at the default of 30 days, Settings > Usage & Spend > Cumulative.

Token activity Cumulative view drawing a running total at a 30 day history window

The card header still reads `Coverage: 30 / 365`, so this is the default window, not the 365-day
workaround. Five columns draw, rising to a full column at the right edge. The five include the week
clipped by the start of the window, which the previous rule discarded.

Before, on `main`, the same view at the same setting is 53 gray columns with no value at any hover.

## Behavior

| | Before | After |
|---|---|---|
| Cumulative columns drawn | 0 of 53 | the scanned window |
| Weekly columns drawn | drops the week clipped by the window | keeps it, for its scanned portion |
| Hover in Cumulative | "Unavailable" everywhere | running totals |
| Leading in-window data gap | unavailable | still unavailable |

## Tests

- `cumulative activity starts at the first scanned week` — an unscanned prefix does not suppress the
  weeks after it.
- `cumulative activity keeps a leading in window gap unavailable` — the regression the review asked
  for. Every week scanned, the first one unresolved, so later totals stay lower bounds.
- `cumulative activity stays unavailable after a gap` — pins the same rule mid-series.
- `series separates an unscanned prefix from an in window gap` — integration level, built through
  `SpendActivitySeries.make` with a 30-day window on the 365-day grid, asserting the scanned window
  draws and that one unresolved in-window day blanks it again.
- `weekly and cumulative activity preserve unavailable coverage` — updated. Its days are all
  scanned, so it now documents the gap case rather than the window-edge case.

## Commands run

- `swift test --filter SpendActivityHeatmapTests` — 19 tests, passed
- `make check` — 0 violations across 1846 files
- `make test` — the full sharded suite passed on the previous revision of this branch and is
  running again against this one locally; CI runs it here as well

## Still open, deliberately

#2893 also notes that only Codex accounts receive the 365-day activity scan, while every other
provider arrives with a snapshot scanned at `costUsageHistoryDays`. The intersection therefore pins
coverage to the smaller window. That is a product decision about scan cost and caching, not a
patch, so it stays out of this PR. With it, the view draws correctly but narrowly at the default
setting.

### CODE REVIEW

File: Sources/CodexBar/SpendActivityHeatmap.swift [hunk 1]
```diff
@@ -129,16 +129,26 @@ struct SpendActivityAggregateSeries: Equatable {
...
         var cumulativeCoverage: [Bool] = []
         for index in self.values.indices {
             total = SpendActivitySeries.saturatingAdd(total, self.values[index])
-            coverageIsComplete = coverageIsComplete && self.isCovered[index]
+            if self.isCovered[index] {
+                hasStarted = true
+            }
+            if hasStarted {
+                coverageIsComplete = coverageIsComplete && self.isCovered[index]
+            }
```
[the remark this answers is not in the document]
Comment (urda) [reply Sources/CodexBar/SpendActivityHeatmap.swift:194]: Good catch, and I agree the case is real.

`isCovered` collapses four different causes into one flag. `TokenActivityInputSummary.tokens(on:)`
returns nil when the day falls outside `coveredInterval`, when the provider is globally invalid,
when the day is in `invalidDays`, and when the day has no entry while `hasCompleteHistory` is
false. Only the first means "never scanned". At this layer the gate cannot tell them apart, so an
unknown day at the leading edge of the scan window is skipped exactly like an unscanned week, and
the weeks after it report complete totals that are really lower bounds.

Separating the two needs the scan window threaded down from `SpendDashboardModel` into
`SpendActivitySeries`, so a day carries three states instead of two: not scanned, scanned but
unknown, covered. That is a larger change, it alters what Weekly draws for the clipped week at the
window edge, and it deserves its own review rather than riding along with a blank-grid fix.

I have added a comment at the gate naming the limitation, and I will note it under the related item
in #2893 so the root cause stays in one thread. Holding it out of this PR deliberately: this change
only decides whether an unscanned week blanks the entire year, and it does not make anything worse
than the current behavior, which shows nothing at all.

**Comment by steipete** [comment]
Thank you @urda — your diagnosis and series-layer design were exactly right, and they landed in https://github.com/steipete/CodexBar/pull/2906 with your commit preserved as author (plus a follow-on hardening for stale-snapshot trailing unscanned days and an end-to-end model regression test). #2893 is fixed on main.

Closing this PR in favor of #2906 since the fork didn't allow maintainer edits. Credit is preserved in the merged commits and the changelog.
